File: typescript/src/ancilis/producers/bedrock.ts
/** BedrockActionProducer — normalizes AWS Bedrock Runtime calls into Actions. */

import { createHash, randomUUID } from "node:crypto";
import type { ResolvedConfig } from "../config/index.js";
import type { Action } from "../engine/action.js";
import { Engine } from "../engine/engine.js";
import { ToolRegistry, ToolStatus } from "../engine/registry.js";
import type { ToolEntry } from "../engine/registry.js";
import type { EvaluationResult } from "../engine/result.js";
import { matchesToolList } from "../engine/tool-matching.js";
import type { EvidenceRecord } from "../evidence/record.js";
import { EvidenceStore } from "../evidence/store.js";
import { canonicalJsonStringify } from "../evidence/chain.js";
import { ProducerType } from "./protocol.js";

const PROVIDER = "aws-bedrock";
const PRODUCER_VERSION = "0.1.0";
const SENSITIVE_KEY_PARTS = [
  "access_key",
  "accesskey",
  "authorization",
  "canonical_request",
  "credential",
  "secret",
  "security_token",
  "session_token",
  "signature",
  "signed_headers",
  "x-amz-security-token",
];
const SAFE_AUTH_MODES = new Set(["iam", "session", "role"]);

type UnknownRecord = Record<string, unknown>;

export interface BedrockInvocation extends UnknownRecord {
  operation?: string;
  operationName?: string;
  operation_name?: string;
  modelId?: string;
  model_id?: string;
  region?: string;
  regionName?: string;
  region_name?: string;
  requestBody?: unknown;
  request_body?: unknown;
  responseBody?: unknown;
  response_body?: unknown;
  streamChunks?: unknown[];
  stream_chunks?: unknown[];
  responseStream?: unknown[];
  response_stream?: unknown[];
  input?: UnknownRecord;
  output?: UnknownRecord;
  response?: UnknownRecord;
  httpStatus?: number;
  http_status?: number;
  requestId?: string;
  request_id?: string;
  latencyMs?: number;
  latency_ms?: number;
  duration_ms?: number;
  headers?: Record<string, unknown>;
  responseMetadata?: UnknownRecord;
  response_metadata?: UnknownRecord;
  agentId?: string;
  agent_id?: string;
  agent?: string;
  agent_name?: string;
  authMode?: string;
  auth_mode?: string;
}

export interface BedrockObservation {
  action: Action;
  evaluation: EvaluationResult;
  evidence: EvidenceRecord;
}

interface NormalizedInvocation {
  operation: string;
  modelId: string;
  region: string | null;
  requestBody: unknown;
  responseBody: unknown;
  streamChunks: unknown[] | null;
  httpStatus: number | null;
  requestId: string | null;
  latencyMs: number | null;
  headers: Record<string, unknown>;
  responseMetadata: Record<string, unknown>;
  agentId: string;
  authMode: string | null;
}

interface ModelMetadata {
  id: string;
  provider: string;
  family: string;
  inference_profile_arn?: string;
}

export class BedrockActionProducer {
  private _config: ResolvedConfig;
  private _engine: Engine;
  private _registry: ToolRegistry;
  private _evidenceStore: EvidenceStore;
  private _sessionId: string = randomUUID();

  constructor(
    config: ResolvedConfig,
    engine: Engine,
    registry?: ToolRegistry,
    evidenceStore?: EvidenceStore,
  ) {
    this._config = config;
    this._engine = engine;
    this._registry = registry ?? engine.registry;
    this._evidenceStore = evidenceStore ?? new EvidenceStore(config);
  }

  get producerType(): ProducerType { return ProducerType.FRAMEWORK; }
  get producerVersion(): string { return PRODUCER_VERSION; }
  get sessionId(): string { return this._sessionId; }

  translate(rawInvocation: unknown): Action {
    const invocation = normalizeInvocation(rawInvocation, this._config.agentName);
    const usage = extractUsage(invocation.responseBody);
    let streamRequestId: string | null = null;
    let streamChunkCount = 0;

    if (invocation.streamChunks) {
      const streamUsage = extractStreamUsage(invocation.streamChunks);
      Object.assign(usage, streamUsage.usage);
      streamRequestId = streamUsage.requestId;
      streamChunkCount = streamUsage.chunkCount;
    }

    const requestId =
      invocation.requestId ??
      streamRequestId ??
      metadataRequestId(invocation.responseMetadata) ??
      headerValue(invocation.headers, "x-amzn-requestid");
    const region = invocation.region ?? regionFromModelId(invocation.modelId);
    const endpoint = endpointFor(region);
    const model = modelMetadata(invocation.modelId);
    const authMode = resolveAuthMode(invocation, invocation.headers);

    const payload: Record<string, unknown> = {
      provider: PROVIDER,
      operation: invocation.operation,
      model_id: invocation.modelId,
      region,
      destination: endpoint,
      http_status: invocation.httpStatus,
      request_id: requestId,
      latency_ms: invocation.latencyMs,
      streaming: invocation.operation === "InvokeModelWithResponseStream" || invocation.streamChunks !== null,
      model,
      deployment: {
        provider: PROVIDER,
        region,
        model_id: invocation.modelId,
        model_family: model.family,
      },
      request: {
        body_present: invocation.requestBody !== undefined && invocation.requestBody !== null,
        body_keys: bodyKeys(invocation.requestBody),
      },
      response: {
        body_present: invocation.responseBody !== undefined && invocation.responseBody !== null,
        body_keys: bodyKeys(invocation.responseBody),
      },
    };
    const deployment = payload["deployment"] as Record<string, unknown>;
    if (model.inference_profile_arn) {
      deployment["inference_profile_arn"] = model.inference_profile_arn;
    }
    if (authMode) payload["auth_mode"] = authMode;
    if (usage.input_tokens !== undefined) payload["input_tokens"] = usage.input_tokens;
    if (usage.output_tokens !== undefined) payload["output_tokens"] = usage.output_tokens;
    if (payload["streaming"] === true) payload["stream"] = { chunk_count: streamChunkCount };

    const toolName = toolNameFor(invocation.operation);
    const entry = this._registry.lookup(toolName);
    const parameterHash = createHash("sha256")
      .update(canonicalJsonStringify(payload))
      .digest("hex");

    return {
      actionId: randomUUID(),
      timestamp: new Date().toISOString(),
      agentId: invocation.agentId,
      sourceType: this.producerType,
      producerType: this.producerType,
      producerVersion: this.producerVersion,
      agentOwner: this._config.agentOwner ?? null,
      actionType: "api_request",
      tool: {
        name: toolName,
        server: endpoint,
        descriptionHash: entry?.descriptionHash ?? null,
      },
      parameters: { raw: payload, parameterHash },
      context: {
        sessionId: this._sessionId,
        dataClassifications: buildDataClassificationCodes(this._config),
        activeOverlays: [...this._config.activeOverlays.keys()],
      },
    };
  }

  async observe(rawInvocation: unknown): Promise<BedrockObservation> {
    const normalized = normalizeInvocation(rawInvocation, this._config.agentName);
    const toolName = this.ensureRegistered(normalized.operation);
    const action = this.translate(rawInvocation);
    const evaluation = this._engine.evaluate(action);
    const evidence = await this._evidenceStore.store(
      evaluation,
      toolName,
      outputSummary(action),
    );
    return { action, evaluation, evidence };
  }

  computeToolHash(toolIdentifier: unknown): string {
    return createHash("sha256").update(String(toolIdentifier)).digest("hex");
  }

  registerTools(registry: ToolRegistry): string[] {
    const registered: string[] = [];
    const now = new Date().toISOString();
    for (const operation of ["InvokeModel", "InvokeModelWithResponseStream"]) {
      const name = toolNameFor(operation);
      registry.register({
        name,
        descriptionHash: this.computeToolHash(name),
        status: ToolStatus.OBSERVED,
        firstSeen: now,
        statusChanged: now,
      } satisfies ToolEntry);
      registered.push(name);
    }
    return registered;
  }

  private ensureRegistered(operation: string): string {
    const name = toolNameFor(operation);
    if (this._registry.lookup(name)) return name;
    const status = matchesToolList(name, this._config.toolsAllowed)
      ? ToolStatus.APPROVED
      : ToolStatus.OBSERVED;
    const now = new Date().toISOString();
    this._registry.register({
      name,
      descriptionHash: this.computeToolHash(name),
      status,
      approvedBy: status === ToolStatus.APPROVED ? "config" : null,
      firstSeen: now,
      statusChanged: now,
    } satisfies ToolEntry);
    return name;
  }
}

export const BedrockAdapter = BedrockActionProducer;

function normalizeInvocation(rawInvocation: unknown, defaultAgentId: string): NormalizedInvocation {
  const raw = isRecord(rawInvocation) ? rawInvocation : {};
  const input = firstRecord(valueAt(raw, "input"));
  const output = firstRecord(valueAt(raw, "output"));
  const response = firstRecord(valueAt(raw, "response"));
  const responseMetadata = firstRecord(
    valueAt(raw, "responseMetadata"),
    valueAt(raw, "response_metadata"),
    valueAt(raw, "ResponseMetadata"),
    valueAt(output, "$metadata"),
    valueAt(response, "$metadata"),
    valueAt(response, "ResponseMetadata"),
  );
  const headers = firstRecord(
    valueAt(raw, "headers"),
    valueAt(raw, "request_headers"),
    valueAt(responseMetadata, "HTTPHeaders"),
    valueAt(responseMetadata, "httpHeaders"),
  );
  const responseBody =
    firstPresent(raw, "responseBody", "response_body") ??
    firstPresent(output, "body", "Body", "responseBody", "response_body") ??
    firstPresent(response, "body", "Body", "responseBody", "response_body");
  const operation =
    optionalString(firstPresent(raw, "operation", "operationName", "operation_name")) ??
    operationFromCommand(rawInvocation) ??
    "InvokeModel";
  const modelId =
    optionalString(firstPresent(raw, "modelId", "model_id", "model")) ??
    optionalString(firstPresent(input, "modelId", "model_id", "model")) ??
    "unknown-model";

  return {
    operation,
    modelId,
    region: optionalString(firstPresent(raw, "region", "regionName", "region_name")),
    requestBody:
      firstPresent(raw, "requestBody", "request_body", "body") ??
      firstPresent(input, "body", "Body", "requestBody", "request_body"),
    responseBody,
    streamChunks: asArray(firstPresent(raw, "streamChunks", "stream_chunks", "responseStream", "response_stream") ?? firstPresent(output, "body")),
    httpStatus:
      optionalNumber(firstPresent(raw, "httpStatus", "http_status", "status_code")) ??
      metadataStatusCode(responseMetadata),
    requestId:
      optionalString(firstPresent(raw, "requestId", "request_id")) ??
      metadataRequestId(responseMetadata),
    latencyMs: optionalNumber(firstPresent(raw, "latencyMs", "latency_ms", "duration_ms")),
    headers,
    responseMetadata,
    agentId:
      optionalString(firstPresent(raw, "agentId", "agent_id", "agent", "agent_name")) ??
      defaultAgentId,
    authMode:
      optionalString(firstPresent(raw, "authMode", "auth_mode")) ??
      nestedAuthMode(raw),
  };
}

function toolNameFor(operation: string): string {
  return `${PROVIDER}:${operation}`;
}

function endpointFor(region: string | null): string {
  return region ? `bedrock-runtime.${region}.amazonaws.com` : "bedrock-runtime.amazonaws.com";
}

function modelMetadata(modelId: string): ModelMetadata {
  const inferenceProfileArn = modelId.includes(":inference-profile/") ? modelId : undefined;
  let modelReference = inferenceProfileArn ? modelId.split("/").pop() ?? modelId : modelId;
  if (modelReference.startsWith("us.")) modelReference = modelReference.slice(3);
  const provider = modelReference.includes(".") ? modelReference.split(".")[0] ?? "unknown" : "unknown";
  let family = "unknown";
  if (modelReference.startsWith("anthropic.claude")) {
    family = "anthropic.claude";
  } else if (modelReference.startsWith("amazon.titan")) {
    family = "amazon.titan";
  } else if (modelReference.includes(".")) {
    family = modelReference.split(".").slice(0, 2).join(".");
  }

  const metadata: ModelMetadata = { id: modelId, provider, family };
  if (inferenceProfileArn) metadata.inference_profile_arn = inferenceProfileArn;
  return metadata;
}

function extractUsage(body: unknown): Record<string, number> {
  const parsed = parseBody(body);
  if (!isRecord(parsed)) return {};

  const usage = valueAt(parsed, "usage");
  if (isRecord(usage)) {
    return tokenUsage(usage, ["input_tokens", "inputTokens", "inputTokenCount"], ["output_tokens", "outputTokens", "outputTokenCount"]);
  }

  const tokenUsageData = tokenUsage(
    parsed,
    ["input_tokens", "inputTokens", "inputTokenCount", "inputTextTokenCount"],
    ["output_tokens", "outputTokens", "outputTokenCount"],
  );
  if (tokenUsageData.output_tokens === undefined) {
    const results = valueAt(parsed, "results");
    if (Array.isArray(results)) {
      let outputTokens = 0;
      let found = false;
      for (const item of results) {
        if (!isRecord(item)) continue;
        const tokenCount = optionalNumber(valueAt(item, "tokenCount"));
        if (tokenCount !== null) {
          outputTokens += tokenCount;
          found = true;
        }
      }
      if (found) tokenUsageData.output_tokens = outputTokens;
    }
  }
  return tokenUsageData;
}

function extractStreamUsage(chunks: unknown[]): {
  usage: Record<string, number>;
  requestId: string | null;
  chunkCount: number;
} {
  const usage: Record<string, number> = {};
  let requestId: string | null = null;
  let chunkCount = 0;
  for (const chunk of chunks) {
    chunkCount += 1;
    const parsedChunk = parseStreamChunk(chunk);
    if (!isRecord(parsedChunk)) continue;
    Object.assign(usage, extractUsage(parsedChunk));
    if (Object.keys(usage).length === 0) {
      const metrics = valueAt(parsedChunk, "amazon-bedrock-invocationMetrics");
      if (isRecord(metrics)) {
        Object.assign(usage, tokenUsage(metrics, ["inputTokenCount", "input_tokens"], ["outputTokenCount", "output_tokens"]));
      }
    }
    const metadata = valueAt(parsedChunk, "metadata");
    if (isRecord(metadata)) {
      Object.assign(usage, extractUsage(metadata));
      requestId ??= optionalString(firstPresent(metadata, "request_id", "requestId", "RequestId"));
    }
  }
  return { usage, requestId, chunkCount };
}

function parseStreamChunk(chunk: unknown): unknown {
  if (isRecord(chunk)) {
    if ("metadata" in chunk) return chunk;
    const chunkPayload = valueAt(chunk, "chunk");
    if (isRecord(chunkPayload) && "bytes" in chunkPayload) {
      return parseBody(valueAt(chunkPayload, "bytes"));
    }
    if ("bytes" in chunk) return parseBody(valueAt(chunk, "bytes"));
  }
  return parseBody(chunk);
}

function parseBody(body: unknown): unknown {
  if (body === null || body === undefined) return null;
  if (body instanceof Uint8Array) {
    try {
      return JSON.parse(Buffer.from(body).toString("utf-8"));
    } catch {
      return null;
    }
  }
  if (isRecord(body)) return body;
  if (typeof body === "string") {
    try {
      return JSON.parse(body);
    } catch {
      return null;
    }
  }
  return null;
}

function bodyKeys(body: unknown): string[] {
  const parsed = parseBody(body);
  if (!isRecord(parsed)) return [];
  return Object.keys(parsed)
    .filter(key => !isSensitiveKey(key))
    .sort();
}

function tokenUsage(data: UnknownRecord, inputKeys: string[], outputKeys: string[]): Record<string, number> {
  const usage: Record<string, number> = {};
  const inputTokens = firstNumber(data, inputKeys);
  const outputTokens = firstNumber(data, outputKeys);
  if (inputTokens !== null) usage["input_tokens"] = inputTokens;
  if (outputTokens !== null) usage["output_tokens"] = outputTokens;
  return usage;
}

function firstNumber(data: UnknownRecord, keys: string[]): number | null {
  for (const key of keys) {
    const value = optionalNumber(valueAt(data, key));
    if (value !== null) return value;
  }
  return null;
}

function resolveAuthMode(invocation: NormalizedInvocation, headers: Record<string, unknown>): string | null {
  if (invocation.authMode) {
    const explicitMode = safeAuthMode(invocation.authMode);
    if (explicitMode) return explicitMode;
  }
  if (headerValue(headers, "x-amz-security-token")) return "session";
  const authorization = headerValue(headers, "authorization");
  if (authorization?.includes("AWS4-HMAC-SHA256")) return "iam";
  return null;
}

function safeAuthMode(value: string): string | null {
  const mode = value.trim().toLowerCase().replaceAll("_", "-");
  return SAFE_AUTH_MODES.has(mode) ? mode : null;
}

function nestedAuthMode(raw: UnknownRecord): string | null {
  const auth = valueAt(raw, "auth");
  if (!isRecord(auth)) return null;
  return optionalString(firstPresent(auth, "mode", "authMode", "auth_mode"));
}

function metadataStatusCode(metadata: UnknownRecord): number | null {
  return optionalNumber(firstPresent(metadata, "HTTPStatusCode", "httpStatusCode", "http_status"));
}

function metadataRequestId(metadata: UnknownRecord): string | null {
  return optionalString(firstPresent(metadata, "RequestId", "requestId", "request_id"));
}

function headerValue(headers: Record<string, unknown>, key: string): string | null {
  for (const [headerKey, value] of Object.entries(headers)) {
    if (headerKey.toLowerCase() === key.toLowerCase()) return String(value);
  }
  return null;
}

function regionFromModelId(modelId: string): string | null {
  if (!modelId.startsWith("arn:")) return null;
  const parts = modelId.split(":");
  return parts.length > 3 && parts[3] ? parts[3] : null;
}

function operationFromCommand(rawInvocation: unknown): string | null {
  if (!rawInvocation || typeof rawInvocation !== "object") return null;
  const ctor = (rawInvocation as { constructor?: { name?: string } }).constructor?.name;
  if (ctor === "InvokeModelCommand" || ctor === "InvokeModelWithResponseStreamCommand") {
    return ctor.replace(/Command$/, "");
  }
  return null;
}

function valueAt(data: unknown, key: string): unknown {
  if (!isRecord(data)) return undefined;
  return data[key];
}

function firstPresent(data: unknown, ...keys: string[]): unknown {
  if (!isRecord(data)) return undefined;
  for (const key of keys) {
    if (Object.prototype.hasOwnProperty.call(data, key)) return data[key];
  }
  return undefined;
}

function firstRecord(...values: unknown[]): Record<string, unknown> {
  for (const value of values) {
    if (isRecord(value)) return { ...value };
  }
  return {};
}

function asArray(value: unknown): unknown[] | null {
  return Array.isArray(value) ? value : null;
}

function optionalString(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  return String(value);
}

function optionalNumber(value: unknown): number | null {
  if (value === null || value === undefined || typeof value === "boolean") return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

function isSensitiveKey(key: string): boolean {
  const lowered = key.toLowerCase().replaceAll("-", "_");
  return SENSITIVE_KEY_PARTS.some(part => lowered.includes(part));
}

function isRecord(value: unknown): value is UnknownRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function buildDataClassificationCodes(config: ResolvedConfig): string[] {
  const codes: string[] = [];
  for (const dcCodes of config.dataClassifications.values()) {
    for (const code of dcCodes) {
      if (!codes.includes(code)) codes.push(code);
    }
  }
  return codes;
}

function outputSummary(action: Action): string {
  const raw = action.parameters.raw;
  return `${raw["provider"]} ${raw["operation"]} ${raw["model_id"]}`;
}

File: python/src/ancilis/producers/bedrock.py
"""AWS Bedrock SDK producer (parity with the TypeScript ``BedrockActionProducer``).

Wraps boto3 ``bedrock-runtime`` calls (``InvokeModel``,
``InvokeModelWithResponseStream``) so each invocation becomes an Action object.
Duck-typed against boto3 — the producer does not import boto3, so the SDK is
not required at runtime.

Mirrors the public API of ``typescript/src/ancilis/producers/bedrock.ts``:
``BedrockInvocation`` / ``BedrockObservation`` / ``BedrockExecutionResult``.
Streaming chunk normalization is intentionally lighter than the TS version;
the common observe→evaluate→execute path is fully covered.
"""

from __future__ import annotations

import functools
import hashlib
import json
import uuid
from collections.abc import Callable, Mapping
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, ParamSpec, TypeVar

from ancilis.config import ResolvedConfig
from ancilis.engine.action import Action, ActionContext, ActionParameters, ToolInfo
from ancilis.engine.engine import Engine
from ancilis.engine.registry import ToolEntry, ToolRegistry, ToolStatus
from ancilis.engine.result import EvaluationResult
from ancilis.evidence.store import EvidenceStore
from ancilis.producers.protocol import ProducerType
from ancilis.producers.tool import BlockedActionError
from ancilis.telemetry import record_adapter_used

P = ParamSpec("P")
R = TypeVar("R")

PROVIDER = "aws-bedrock"
PRODUCER_VERSION = "0.1.0"
DEFAULT_OPERATION = "InvokeModel"
STREAM_OPERATION = "InvokeModelWithResponseStream"


@dataclass
class BedrockInvocation:
    """Provider-agnostic shape covering the boto3 bedrock-runtime call surface."""

    operation: str = DEFAULT_OPERATION
    model_id: str = "unknown-model"
    region: str | None = None
    agent_id: str | None = None
    request_body: Any = None
    response_body: Any = None
    stream_chunks: list[Any] | None = None
    http_status: int | None = None
    request_id: str | None = None
    latency_ms: float | None = None
    headers: dict[str, Any] = field(default_factory=dict)
    response_metadata: dict[str, Any] = field(default_factory=dict)
    auth_mode: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class BedrockObservation:
    action: Action
    evaluation: EvaluationResult


@dataclass
class BedrockExecutionResult:
    action: Action
    evaluation: EvaluationResult
    blocked: bool
    response: Any = None


def _first_present(mapping: Mapping[str, Any] | None, *keys: str) -> Any:
    if not mapping:
        return None
    for key in keys:
        if key in mapping and mapping[key] is not None:
            return mapping[key]
    return None


def _model_metadata(model_id: str) -> dict[str, str]:
    """Derive provider/family from a Bedrock model id or inference-profile ARN."""
    inference_profile_arn = model_id if ":inference-profile/" in model_id else None
    model_reference = (
        model_id.rsplit("/", 1)[-1] if inference_profile_arn else model_id
    )
    if model_reference.startswith("us."):
        model_reference = model_reference[3:]
    provider = model_reference.split(".", 1)[0] if "." in model_reference else "unknown"
    if model_reference.startswith("anthropic.claude"):
        family = "anthropic.claude"
    elif model_reference.startswith("amazon.titan"):
        family = "amazon.titan"
    elif "." in model_reference:
        parts = model_reference.split(".")
        family = ".".join(parts[:2])
    else:
        family = "unknown"
    out = {"id": model_id, "provider": provider, "family": family}
    if inference_profile_arn:
        out["inference_profile_arn"] = inference_profile_arn
    return out


def _endpoint_for(region: str | None) -> str:
    return f"bedrock-runtime.{region}.amazonaws.com" if region else "bedrock-runtime.amazonaws.com"


def _body_keys(body: Any) -> list[str]:
    if isinstance(body, Mapping):
        return sorted(body.keys())
    if isinstance(body, str):
        try:
            parsed = json.loads(body)
        except (json.JSONDecodeError, ValueError):
            return []
        if isinstance(parsed, Mapping):
            return sorted(parsed.keys())
    return []


def _extract_usage(body: Any) -> dict[str, int]:
    parsed: Any = body
    if isinstance(body, str):
        try:
            parsed = json.loads(body)
        except (json.JSONDecodeError, ValueError):
            return {}
    if not isinstance(parsed, Mapping):
        return {}
    usage_candidate = parsed.get("usage")
    usage_obj: Mapping[str, Any] = (
        usage_candidate if isinstance(usage_candidate, Mapping) else parsed
    )
    out: dict[str, int] = {}
    for key, aliases in (
        ("input_tokens", ("input_tokens", "inputTokens", "inputTokenCount", "inputTextTokenCount")),
        ("output_tokens", ("output_tokens", "outputTokens", "outputTokenCount")),
    ):
        for alias in aliases:
            if alias in usage_obj and isinstance(usage_obj[alias], (int, float)):
                out[key] = int(usage_obj[alias])
                break
    return out


def _normalize_invocation(
    raw: BedrockInvocation | Mapping[str, Any] | None, default_agent_id: str
) -> BedrockInvocation:
    if isinstance(raw, BedrockInvocation):
        if not raw.agent_id:
            raw.agent_id = default_agent_id
        if not raw.operation:
            raw.operation = DEFAULT_OPERATION
        if not raw.model_id:
            raw.model_id = "unknown-model"
        return raw
    raw_map: Mapping[str, Any] = raw if isinstance(raw, Mapping) else {}
    response_metadata = _first_present(
        raw_map, "responseMetadata", "response_metadata", "ResponseMetadata"
    )
    if not isinstance(response_metadata, Mapping):
        response_metadata = {}
    headers = _first_present(raw_map, "headers", "request_headers")
    if not isinstance(headers, Mapping):
        headers = response_metadata.get("HTTPHeaders") or response_metadata.get("httpHeaders") or {}
    operation = _first_present(raw_map, "operation", "operationName", "operation_name") or DEFAULT_OPERATION
    model_id = _first_present(raw_map, "modelId", "model_id", "model") or "unknown-model"
    request_id = _first_present(raw_map, "requestId", "request_id")
    if request_id is None and isinstance(response_metadata, Mapping):
        request_id = response_metadata.get("RequestId") or response_metadata.get("requestId")
    http_status = _first_present(raw_map, "httpStatus", "http_status", "status_code")
    if http_status is None and isinstance(response_metadata, Mapping):
        http_status = response_metadata.get("HTTPStatusCode") or response_metadata.get("httpStatusCode")
    return BedrockInvocation(
        operation=str(operation),
        model_id=str(model_id),
        region=_first_present(raw_map, "region", "regionName", "region_name"),
        agent_id=_first_present(raw_map, "agentId", "agent_id", "agent", "agent_name") or default_agent_id,
        request_body=_first_present(raw_map, "requestBody", "request_body", "body"),
        response_body=_first_present(raw_map, "responseBody", "response_body"),
        stream_chunks=_first_present(raw_map, "streamChunks", "stream_chunks", "responseStream", "response_stream"),
        http_status=int(http_status) if isinstance(http_status, (int, float)) else None,
        request_id=str(request_id) if request_id else None,
        latency_ms=_first_present(raw_map, "latencyMs", "latency_ms", "duration_ms"),
        headers=dict(headers) if isinstance(headers, Mapping) else {},
        response_metadata=dict(response_metadata) if isinstance(response_metadata, Mapping) else {},
        auth_mode=_first_present(raw_map, "authMode", "auth_mode"),
        metadata={
            k: v
            for k, v in raw_map.items()
            if k not in {
                "operation", "operationName", "operation_name",
                "modelId", "model_id", "model",
                "region", "regionName", "region_name",
                "agentId", "agent_id", "agent", "agent_name",
                "requestBody", "request_body", "body",
                "responseBody", "response_body",
                "streamChunks", "stream_chunks", "responseStream", "response_stream",
                "httpStatus", "http_status", "status_code",
                "requestId", "request_id",
                "latencyMs", "latency_ms", "duration_ms",
                "headers", "request_headers",
                "responseMetadata", "response_metadata", "ResponseMetadata",
                "authMode", "auth_mode",
            }
        },
    )


def _tool_name_for(operation: str) -> str:
    return f"{PROVIDER}:{operation}"


class BedrockActionProducer:
    """Produces Action objects for AWS Bedrock SDK calls.

    Pairs with the TypeScript ``BedrockActionProducer``. Use ``observe`` for
    a fully-formed invocation already collected from the SDK; ``execute`` to
    wrap a transport callable; ``wrap_invoke_model`` to wrap a boto3 client's
    ``invoke_model`` method.
    """

    def __init__(
        self,
        config: ResolvedConfig,
        engine: Engine,
        registry: ToolRegistry | None = None,
        evidence_store: EvidenceStore | None = None,
    ) -> None:
        self._config = config
        self._engine = engine
        self._registry = registry or engine.registry
        self._evidence_store = evidence_store if evidence_store is not None else EvidenceStore(config)
        self._session_id = str(uuid.uuid4())
        record_adapter_used(PROVIDER)

    @property
    def session_id(self) -> str:
        return self._session_id

    @property
    def producer_type(self) -> ProducerType:
        return ProducerType.FRAMEWORK

    @property
    def producer_version(self) -> str:
        return PRODUCER_VERSION

    def translate(self, raw_invocation: BedrockInvocation | Mapping[str, Any]) -> Action:
        invocation = _normalize_invocation(raw_invocation, self._config.agent_name)
        usage = _extract_usage(invocation.response_body)
        endpoint = _endpoint_for(invocation.region)
        model = _model_metadata(invocation.model_id)
        is_streaming = invocation.operation == STREAM_OPERATION or invocation.stream_chunks is not None

        payload: dict[str, Any] = {
            "provider": PROVIDER,
            "operation": invocation.operation,
            "model_id": invocation.model_id,
            "region": invocation.region,
            "destination": endpoint,
            "http_status": invocation.http_status,
            "request_id": invocation.request_id,
            "latency_ms": invocation.latency_ms,
            "streaming": is_streaming,
            "model": model,
            "deployment": {
                "provider": PROVIDER,
                "region": invocation.region,
                "model_id": invocation.model_id,
                "model_family": model["family"],
            },
            "request": {
                "body_present": invocation.request_body is not None,
                "body_keys": _body_keys(invocation.request_body),
            },
            "response": {
                "body_present": invocation.response_body is not None,
                "body_keys": _body_keys(invocation.response_body),
            },
        }
        if "inference_profile_arn" in model:
            payload["deployment"]["inference_profile_arn"] = model["inference_profile_arn"]
        if invocation.auth_mode:
            payload["auth_mode"] = invocation.auth_mode
        if "input_tokens" in usage:
            payload["input_tokens"] = usage["input_tokens"]
        if "output_tokens" in usage:
            payload["output_tokens"] = usage["output_tokens"]
        if is_streaming and invocation.stream_chunks is not None:
            payload["stream"] = {"chunk_count": len(invocation.stream_chunks)}

        tool_name = _tool_name_for(invocation.operation)
        entry = self._registry.lookup(tool_name)
        param_hash = hashlib.sha256(
            json.dumps(payload, sort_keys=True, default=repr).encode()
        ).hexdigest()
        dc_codes: list[str] = []
        for codes in self._config.data_classifications.values():
            for code in codes:
                if code not in dc_codes:
                    dc_codes.append(code)

        return Action(
            action_id=str(uuid.uuid4()),
            timestamp=datetime.now(timezone.utc).isoformat(),
            agent_id=invocation.agent_id or self._config.agent_name,
            source_type=self.producer_type.value,
            agent_owner=self._config.agent_owner or None,
            action_type="api_request",
            tool=ToolInfo(
                name=tool_name,
                server=endpoint,
                description_hash=entry.description_hash if entry else None,
            ),
            parameters=ActionParameters(raw=payload, parameter_hash=param_hash),
            context=ActionContext(
                data_classifications=dc_codes,
                active_overlays=list(self._config.active_overlays.keys()),
            ),
            producer_type=self.producer_type.value,
            producer_version=self.producer_version,
        )

    def compute_tool_hash(self, tool_identifier: Any) -> str:
        return hashlib.sha256(str(tool_identifier).encode()).hexdigest()

    def register_tools(self, registry: ToolRegistry) -> list[str]:
        registered: list[str] = []
        for operation in (DEFAULT_OPERATION, STREAM_OPERATION):
            name = _tool_name_for(operation)
            if registry.lookup(name) is None:
                registry.register(
                    ToolEntry(
                        name=name,
                        description_hash=self.compute_tool_hash(name),
                        status=ToolStatus.OBSERVED,
                    )
                )
            registered.append(name)
        return registered

    def _ensure_registered(self, operation: str) -> str:
        name = _tool_name_for(operation)
        if self._registry.lookup(name) is None:
            status = (
                ToolStatus.APPROVED
                if name in self._config.tools_allowed
                else ToolStatus.OBSERVED
            )
            self._registry.register(
                ToolEntry(
                    name=name,
                    description_hash=self.compute_tool_hash(name),
                    status=status,
                    approved_by="config" if status == ToolStatus.APPROVED else None,
                )
            )
        return name

    def observe(
        self, raw_invocation: BedrockInvocation | Mapping[str, Any]
    ) -> BedrockObservation:
        invocation = _normalize_invocation(raw_invocation, self._config.agent_name)
        tool_name = self._ensure_registered(invocation.operation)
        action = self.translate(invocation)
        evaluation = self._engine.evaluate(action)
        self._evidence_store.store(evaluation, tool_name=tool_name)
        return BedrockObservation(action=action, evaluation=evaluation)

    def execute(
        self,
        raw_invocation: BedrockInvocation | Mapping[str, Any],
        *,
        transport: Callable[P, R],
        transport_args: tuple[Any, ...] = (),
        transport_kwargs: dict[str, Any] | None = None,
        enforce: bool = False,
    ) -> BedrockExecutionResult:
        observation = self.observe(raw_invocation)
        if enforce and observation.evaluation.decision == "BLOCK":
            raise BlockedActionError(observation.action.tool.name, observation.evaluation)
        response = transport(*transport_args, **(transport_kwargs or {}))
        return BedrockExecutionResult(
            action=observation.action,
            evaluation=observation.evaluation,
            blocked=False,
            response=response,
        )

    def wrap_invoke_model(
        self,
        invoke_model: Callable[..., Any],
        *,
        agent_name: str | None = None,
        operation: str = DEFAULT_OPERATION,
        enforce: bool = False,
    ) -> Callable[..., BedrockExecutionResult]:
        """Wrap a boto3 ``client.invoke_model`` so each call is observed first."""

        @functools.wraps(invoke_model)
        def wrapped(*args: Any, **kwargs: Any) -> BedrockExecutionResult:
            invocation = BedrockInvocation(
                operation=operation,
                model_id=str(kwargs.get("modelId") or kwargs.get("model_id") or "unknown-model"),
                agent_id=agent_name or self._config.agent_name,
                request_body=kwargs.get("body"),
            )
            return self.execute(
                invocation,
                transport=invoke_model,
                transport_args=args,
                transport_kwargs=kwargs,
                enforce=enforce,
            )

        return wrapped


# Alias matches the TS export.
BedrockAdapter = BedrockActionProducer
